src/main/java/io/pravega/leaderelection/LeaderElection.java

```diff
@@ -0,0 +1,493 @@
+        }
+
+        private static int compare(Entry<String, InstanceInfo> o1 , Entry<String, InstanceInfo> o2) {
+            return Long.compare(o1.getValue().times, o2.getValue().times);
+        }
+
+        /**
+         * Return all instances that are dead at given time.
+         * @param vectorTime The given time.
+         * @return A list of the instances that are dead.
+         */
+        private List<String> findInstancesThatWillDieBy(long vectorTime) {
+            double deathThreshold;
+            double deathCycle;
+            List<String> res = new ArrayList<>();
+            for (String key: liveInstances.keySet()) {
+                deathCycle = DEATH_THRESHOLD * liveInstances.get(key).timeout;
+                deathThreshold = (deathCycle + 1) * (liveInstances.size() - 1);
+                if (liveInstances.get(key).timestamp < (vectorTime - deathThreshold)) {
+                    res.add(key);
+                }
+            }
+            return res;
+        }
+
+        /**
+         * Check if the instance is healthy or not.
+         * @param name The name of the instance.
+         * @return TRUE means the instance is healthy, FALSE means not healthy.
+         */
+        private boolean isHealthy(String name) {
+            long unhealthyThreshold = vectorTime - UNHEALTHY_THRESHOLD * liveInstances.size();
+            InstanceInfo instanceinfo = liveInstances.get(name);
+            return instanceinfo == null || instanceinfo.timestamp >= unhealthyThreshold;
+        }
+        /**
+         * Return all instances that are alive.
+         * @return A set of alive instances.
+         */
+        private Set<String> getLiveInstances() {
+            return Collections.unmodifiableSet(liveInstances.keySet());
+        }
+
+    }
+
+    @RequiredArgsConstructor
+    private static class CreateState implements Serializable, InitialUpdate<LiveInstances> {
+        private static final long serialVersionID = 1L;
+        private final Map<String, InstanceInfo> liveInstances;
+
+        @Override
+        public LiveInstances create(String scopedStreamName, Revision revision) {
+            return new LiveInstances(scopedStreamName, revision, liveInstances, 0, null);
+        }
+    }
+
+    private static abstract class HeartbeatUpdate implements Update<LiveInstances>, Serializable {
+        private static final long serialVersionUID = 1L;
+    }
+
+    @RequiredArgsConstructor
+    private static class HeartBeat extends HeartbeatUpdate {
+        private static final long serialVersionUID = 1L;
+        private final String name;
+        private final InstanceInfo instanceInfo;
+
+        @Override
+        public LiveInstances applyTo(LiveInstances state, Revision newRevision) {
+            Map<String, InstanceInfo> tempInstances = new HashMap<>(state.liveInstances);
+            long vectorTime = Long.max(tempInstances.values().stream().max(InstanceInfo::compare).get().timestamp, instanceInfo.timestamp);
```

Comment:
This could be done more cleanly by first mapping the instance to the timestamp. 
Then you could use the builtin for Max, and not need the compare and the result would be the timestamp.